flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/AsyncExecutionController.java

```diff
@@ -0,0 +1,241 @@
+    }
+
+    /**
+     * Each time before a code segment (callback) is about to run in mailbox (task thread), this
+     * method should be called to switch a context in AEC.
+     *
+     * @param switchingContext the context to switch.
+     */
+    public void setCurrentContext(RecordContext<R, K> switchingContext) {
+        currentContext = switchingContext;
+    }
+
+    /**
+     * Submit a {@link StateRequest} to this AEC and trigger if needed.
+     *
+     * <p>This method should only be invoked in task-thread based on the assumption of all callbacks
+     * should be executed in mailbox. Please make this method thread-safe if the assumption changes.
+     *
+     * @param state the state to request.
+     * @param type the type of this request.
+     * @param payload the payload input for this request.
+     * @return the state future.
+     */
+    public <IN, OUT> InternalStateFuture<OUT> handleRequest(
+            @Nullable State state, StateRequest.RequestType type, @Nullable IN payload) {
+        // Step 1: build state future & assign context.
+        InternalStateFuture<OUT> stateFuture = stateFutureFactory.build(currentContext);
+
+        StateRequest<K, IN, OUT> request =
+                new StateRequest<>(state, type, payload, stateFuture, currentContext);
+
+        // Step 2: try to seize the capacity, if the current in-flight records exceeds the limit,
+        // block the current state request from entering until some buffered requests are processed.
+        seizeCapacity();
+
+        // Step 3: try to occupy the key and place it into right buffer.
+        if (currentContext.tryOccupyKey()) {
+            insertActiveBuffer(request);
+        } else {
+            insertBlockingBuffer(request);
+        }
+        // Step 4: trigger the (active) buffer if needed.
+        triggerIfNeeded(false);
+        return stateFuture;
+    }
+
+    <IN, OUT> void insertActiveBuffer(StateRequest<K, IN, OUT> request) {
+        activeBuffer.add(request);
+    }
+
+    <IN, OUT> void insertBlockingBuffer(StateRequest<K, IN, OUT> request) {
+        blockingBuffer.add(request);
+    }
+
+    /**
+     * Trigger a batch of requests.
+     *
+     * @param force whether to trigger requests in force.
+     */
+    @VisibleForTesting
+    void triggerIfNeeded(boolean force) {
+        if (!force && activeBuffer.size() < batchSize) {
+            return;
+        }
+        migrateBlockingToActive();
+        List<StateRequest<?, ?, ?>> toRun = new LinkedList<>(activeBuffer);
+        activeBuffer.clear();
+        stateExecutor.executeBatchRequests(toRun);
+        // migrate state requests in blocking buffer to active buffer actively, let the requests
+        // that enter the buffer first be processed first.
+        migrateBlockingToActive();
+    }
+
+    /** Migrate the blocking requests to the active buffer. */
+    @VisibleForTesting
+    void migrateBlockingToActive() {
```

Comment:
Personally I'd suggest a queueing mechanism, meaning that each `blocked request` will queue under one key. While the `active request/context`'s reference counting reached zero, it pop one request from the corresponding queue of that key and put that into `active buffer`. I think this is a high-performance implementation.

Reply:
Thanks for the suggestion, I add a new class `StateRequestsBuffer` which groups state requests in the blocking buffer by key. Migrating one state request from blocking buffer to active buffer will trigger in `RecordContext#disposer`.